src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -925,301 +1296,410 @@ internal TransitionRegex<TSet> CreateDerivative()
-            // stack first end up at the tail of the concatenation.
-            static SymbolicRegexNode<TSet> BuildLeaf(SymbolicRegexNode<TSet> head, List<SymbolicRegexNode<TSet>> continuation)
+            //caching pruning to avoid otherwise potential quadratic worst case behavior
+            SymbolicRegexNode<TSet>? prunedNode;
+            (SymbolicRegexNode<TSet>, uint) key = (this, context);
+            if (_builder._pruneCache.TryGetValue(key, out prunedNode))
             {
-                SymbolicRegexNode<TSet> leaf = head._builder.Epsilon;
-                for (int i = 0; i < continuation.Count; ++i)
-                {
-                    leaf = head._builder.CreateConcat(continuation[i], leaf);
-                }
-                return head._builder.CreateConcat(head, leaf);
+                return prunedNode;
             }
 
-            // Helper function for detecting when an unconditionally nullable pattern is in the transitions and no
-            // more transitions need to be considered. If the derivative simulates backtracking, then in the next
-            // step nothing after an unconditionally nullable state would be considered.
-            // For example, d_a( a|ab ) under backtracking simulation transitions to just epsilon, while without
-            // backtracking simulation it would transition to epsilon|b.
-            // All parts of the function below should consider IsDone and return early when it is true.
-            static bool IsDone(TransitionList transitions, bool simulateBacktracking) =>
-                simulateBacktracking && transitions.Count > 0 && transitions.LastTransition.Node.IsNullable;
-
-            // Nothing and epsilon can't consume a character so they generate no transition
-            if (IsNothing || IsEpsilon)
+            if (!StackHelper.TryEnsureSufficientExecutionStack())
             {
-                return;
+                return StackHelper.CallOnEmptyStack(PruneLowerPriorityThanNullability, context);
             }
 
-            // For both .* and .+ the derivative is .* for any character
-            if (IsAnyStar || IsAnyPlus)
+            switch (_kind)
             {
-                Debug.Assert(!IsDone(transitions, simulateBacktracking));
-                SymbolicRegexNode<TSet> leaf = BuildLeaf(_builder._anyStar, continuation);
-                transitions.Add(leaf, effects?.ToArray() ?? Array.Empty<DerivativeEffect>());
-                // Signal early exit if the leaf is unconditionally nullable
-                return;
+                case SymbolicRegexNodeKind.OrderedOr:
+                    Debug.Assert(_left is not null && _right is not null);
+                    // The left alternative, when nullable, has priority over the right alternative
+                    // Otherwise the left alternative is still active and the right alternative is pruned
```

Comment:
I'm not clear on what pruning means in this context.  We still need to keep the branch, it just doesn't have priority, right?

Reply:
In a situation (X|Y) where X is nullable (in the given context), Y is never used, thus omitted (forgotten).
Sort of like taking the higher-priority branch in backtracking that is known to lead to a match at which point the other branches become irrelevant and are no  longer followed.

Reply:
updated the comment in the code accordingly:
```
// In a alternation (X|Y) where X is nullable (in the given context), Y must be eliminated.
// Thus, taking the higher-priority branch in backtracking that is known to lead to a match
// at which point the other branches become irrelevant and must no longer be used.
```